feat: add mysql sample

jdpedrie opened the pull request:
Closes #991.

cc @kurtisvg.

kurtisvg commented on cloud_sql/mysql/pdo/src/DB.php:
```diff
@@ -0,0 +1,54 @@
+        // // $schema = 'your_db_name';
+        // // $cloud_sql_connection_name = getenv("CLOUD_SQL_CONNECTION_NAME");
+
+        if ($cloud_sql_connection_name) {
+            // Connect using UNIX sockets
+            $dsn = sprintf(
+                'mysql:dbname=%s;unix_socket=/cloudsql/%s',
+                $schema,
+                $cloud_sql_connection_name
+            );
+        } else {
+            // Connect using TCP
+            // $hostname = '127.0.0.1';
+            $dsn = sprintf('mysql:dbname=%s;host=%s', $schema, $hostname);
+        }
```
@shubha-rajan - I'll defer to you if we want to figure out how to handle split TCP/Unix socket connections snippets in this PR or fix all the PHP samples together later.

shubha-rajan replied:
I think we should figure out how to handle TCP/Unix socket connections and update the postgres sample for consistency as soon as this is merged.  One note I have right now is that if we're checking for the existence of `CLOUD_SQL_CONNECTION_NAME` to decide whether to use TCP or Unix Sockets, we should add a note in the readme to not set that env variable if they want to use TCP.

grayside commented on cloud_sql/mysql/pdo/index.php:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    echo $message;
+}
```
This is missing 404/405 route error handling, so any other requests sent to the service will receive a 200 OK.

Something like the following:

```php
elseif ($_SERVER['REQUEST_URI'] == '/') {
  return http_response_code(405);
}
else {
  return http_response_code(404);
}
```

jdpedrie commented:
@grayside @kurtisvg would you like me to apply all these suggestions to the [postgres](https://github.com/GoogleCloudPlatform/php-docs-samples/tree/master/cloud_sql/postgres/pdo) samples as well, which provided the template for this and the sqlserver PRs?

kurtisvg commented:
> @grayside @kurtisvg would you like me to apply all these suggestions to the [postgres](https://github.com/GoogleCloudPlatform/php-docs-samples/tree/master/cloud_sql/postgres/pdo) samples as well, which provided the template for this and the sqlserver PRs?

Yes please!

shubha-rajan commented on cloud_sql/mysql/pdo/README.md:
````diff
@@ -0,0 +1,66 @@
+export DB_HOSTNAME='localhost'
+```
+
+Note: Saving credentials in environment variables is convenient, but not secure - consider a more secure solution such as [Cloud KMS](https://cloud.google.com/kms/) to help keep secrets safe.
````
Nit: We're moving towards suggesting [Secret Manager](https://cloud.google.com/secret-manager/docs/overview) instead of KMS.

jdpedrie commented:
@grayside @shubha-rajan @kurtisvg @bshaffer I refactored the app to use the Slim 4 framework used [elsewhere](https://github.com/GoogleCloudPlatform/php-docs-samples/tree/master/appengine/php72/cloudsql) in the PHP docs samples. I believe I've addressed all the comments, but I haven't written any tests yet. I'd like to make sure you are all in favor of the new direction before I wrap it up. Once everyone is happy, I'll also refactor the sqlserver and postgres samples in a similar direction.

jdpedrie commented on cloud_sql/mysql/pdo/README.md:
```diff
@@ -0,0 +1,72 @@
+
+Navigate towards http://localhost:8080 to verify your application is running correctly.
+
+## Google App Engine Standard
```
(todo https://github.com/GoogleCloudPlatform/php-docs-samples/pull/1053#discussion_r408338349 )

grayside commented on cloud_sql/mysql/pdo/src/app.php:
```diff
@@ -0,0 +1,94 @@
+            $dsn = sprintf('mysql:dbname=%s;host=%s', $schema, $hostname);
+        }
+
+        $conn = new PDO($dsn, $username, $password);
```
In the event this fails on something other than a PDOException, this is pretty messy:

<img width="1372" alt="Screen Shot 2020-04-23 at 9 03 27 PM" src="https://user-images.githubusercontent.com/283489/80173878-01d4a580-85a6-11ea-9fc2-6ce0b30aa1f0.png">

The problem in this case is a TypeError because I didn't configure the environment values. We could catch a TypeError and say "wake up and pay attention to the env values", but maybe a more generic alternative to `PDOException` in `Exception` would be appropriate as a catch-all? (The risk of this is more pronounced in local development or in a future adaptation of this sample for Cloud Run.)

Configuring with non-working values was successfully caught by the exception handler.

jdpedrie replied:
I've added a TypeError catch to improve the error message users will get.

grayside commented on cloud_sql/mysql/pdo/index.php:
```diff
@@ -0,0 +1,50 @@
+$app = include __DIR__ . '/src/app.php';
+
+$app->get('/', function ($request, $response) {
+    return $this->get('view')->render($response, 'template.twig', [
```
I like this pattern of thin controller and nicely separated model and view :)

What happens if the twig template fails to compile (e.g., syntax error) or a query fails?

jdpedrie replied:
If it raises an exception, a similar error handling screen to the Slim one elsewhere.

grayside commented on cloud_sql/mysql/pdo/views/template.twig:
```diff
@@ -0,0 +1,99 @@
+    </div>
+    <h4 class="header center">Recent Votes</h4>
+    <ul class="container collection center">
+        {% for vote in votes %}
```
nit: If there are no votes yet is there something different we want to do?

jdpedrie replied:
Good idea! I added an if/else and a "No votes have been cast" message.

jdpedrie commented on cloud_sql/mysql/pdo/src/app.php:
```diff
@@ -0,0 +1,94 @@
+            );
+        } else {
+            // Connect using TCP
+            // $hostname = '127.0.0.1';
```
I removed it here and added it to the list of commented-out variables above with a note advising users to only set it in TCP mode.

kurtisvg commented on cloud_sql/mysql/pdo/views/template.twig:
```diff
@@ -0,0 +1,99 @@
+</div>
+<script>
+    function vote(voteValue) {
+        var xhr = new XMLHttpRequest();
```
/cc @shubha-rajan - might be good to add to our list of updates

shubha-rajan commented on cloud_sql/mysql/pdo/src/app.php:
```diff
@@ -0,0 +1,111 @@
+            $dsn = sprintf('mysql:dbname=%s;host=%s', $dbName, $hostname);
+        }
+
+        $conn = new PDO($dsn, $username, $password);
```
Could this be configured to set a connection timeout? It might be a good idea to show an example of that in the sample.

jdpedrie replied:
Sure! I added connection timeout and also demonstrated how to set PDO to throw exceptions by default if any errors occur.

shubha-rajan commented on cloud_sql/mysql/pdo/src/app.php:
```diff
@@ -0,0 +1,111 @@
+    $cloud_sql_connection_name = getenv('CLOUD_SQL_CONNECTION_NAME');
+
+    try {
+        // # [START cloud_sql_mysql_pdo_create]
```
The other languages' samples on the Managing Connections page where this sample will be published implement connection pools, but from quickly researching (I'm not too familiar with PHP), there isn't really a way to use pooling with PHP/PDO. Can you include a comment explaining how PDO handles connections and why we're not using pooling here, to make it less confusing in the context of the page? (https://cloud.google.com/sql/docs/mysql/manage-connections)

jdpedrie replied:
Some PDO drivers support connection pooling ([sqlserver](https://docs.microsoft.com/en-us/sql/connect/php/connection-pooling-microsoft-drivers-for-php-for-sql-server?view=sql-server-ver15) for instance) via driver configuration (not configured at the application level).

In general usage it's a bit less of a concern for PHP due to the manner in which PHP lifecycles work. PHP applications are usually not long running, they start at the beginning of request processing and more or less shut down at the end (this is a bit tricky because certain runtimes like php-fpm don't fully shut down). Even if they don't fully shut down, all application state is lost, including database connections.

Event loops are becoming more common in PHP, and that's a case where connection pooling begins to be more of a requirement. In the case of PDO drivers which don't support pooling, I've seen cases where the application will simply manage a pool of `PDO` instances. `new PDO()` connects immediately, at least for mysql. If anything is mentioned, it should be this option, I believe.

The other alternative is persistent connections, but the pitfalls there are rather dangerous and it's not something I think we should recommend.

shubha-rajan commented on cloud_sql/mysql/pdo/src/Votes.php:
```diff
@@ -0,0 +1,103 @@
+     * @param string $value The value to vote for.
+     * @return boolean
+     */
+    public function insertVote(string $value) : bool
+    {
+        $sql = "INSERT INTO votes (time_cast, vote_value) VALUES (NOW(), :voteValue)";
+        $statement = $this->connection->prepare($sql);
+        $statement->bindParam('voteValue', $value);
+
+        return $statement->execute();
+    }
```
This might need to be wrapped in `cloud_sql_mysql_pdo_connection` region tags with a comment about how PDO handles opening and closing connections. 
See the python sample here for an example: https://cloud.google.com/sql/docs/mysql/manage-connections#open-close-python

shubha-rajan replied:
That looks great. Thank you.

jdpedrie commented:
@grayside could you please take another look?

grayside commented on cloud_sql/mysql/pdo/src/app.php:
```diff
@@ -0,0 +1,116 @@
+            PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION
+        ]);
+        # [END cloud_sql_mysql_pdo_create]
+    } catch (TypeError $e) {
```
I'm loving the troubleshooting guidance in this error handling.

shubha-rajan commented:
@jdpedrie looks like a checkstyle issue is causing the build to fail.  I can merge this as soon as that's fixed and the tests pass.

![image](https://user-images.githubusercontent.com/19631367/81095042-453de680-8eb9-11ea-8515-a91a46bc888a.png)
